### PR TITLE
feature/33-add-chat-data-classes

### LINKED ISSUE #33: Add Chat Data Classes
**Describe the solution you'd like**

Add data classes for the chat system.
Add javadoc to the chat data classes.

**Additional Comments**

To be stored in the database:
- existing channels
- chat history

You can find some information about the chat system in the [old backend](https://github.com/cryptic-game/cryptic-chat/tree/issue/1-create-chat-microservice).

Some examples of javadoc can be found in data classes with existing javadoc.

### PR DESCRIPTION
**Pull request by bene99**
**Description**
implemented the data classes for the chat system
added Java Data Classes: Channel, Channel Access, Message
added enum ChatAction to classify the type of Message
(e.g. whisper-msg, send-message(normal msg), member-join...)

Closes #33 

**Testing Instructions**
1. Testing the channels
- call create channel to create a channel & see if an entry in the table is added
- call delete channel to delete a channel & see if the entry in the table is deleted
2. Testing the ChannelAccess Class
- test if users can be added to and deleted from a channel by calling ChannelAccess.join and
ChannelAccess.remove
- test if ChannelAccess.getMembers returns all memebs of a channel
3. Test the Message Class
- test the function Message.send() with different types of messages, like a normal message, a whisper, a member joining
- test if the functionMessage.getMessages() returns 
1. all messages sent to all users 
2. the whispered messages form or to the user

**Additional Notes**
/

### CODE REVIEW

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Channel channel = (Channel) o;
+        return Objects.equals(getName(), channel.getName());
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java:63]: Add UUID checking

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getName());
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java:68]: Add UUID to Hash

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    public static boolean leave(User user, Channel channel) {
+        try (Session sqlSession = sqlConnection.openSession()) {
+            sqlSession
+                    .createQuery("delete from ChannelAccess ca where ca.channel = :channel and ca.user = :user")
+                    .setParameter("channel", channel)
+                    .setParameter("user", user);
+            return true;
+        } catch (HibernateException e) {
+            return false;
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:70]: Log Error

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ChannelAccess that = (ChannelAccess) o;
+        return Objects.equals(getUser(), that.getUser()) &&
+                Objects.equals(getChannel(), that.getChannel());
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:90]: Add UUID checking

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getUser(), getChannel());
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:95]: Add UUID to Hash

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        message.setTarget(target);
+        message.setType(type);
+        message.setText(text);
+
+        sqlSession.beginTransaction();
+        sqlSession.save(message);
+        sqlSession.getTransaction().commit();
+        sqlSession.close();
+        return message;
+    }
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:106]: Notify All Users in the Channel

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        ChannelAccess channelAccess = new ChannelAccess();
+        channelAccess.setUser(user);
+        channelAccess.setChannel(channel);
+
+        sqlSession.beginTransaction();
+        sqlSession.save(channelAccess);
+        sqlSession.getTransaction().commit();
+        sqlSession.close();
+        return channelAccess;
+    }
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:60]: Notifiy Users

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java (hunk 3 above)
Comment (MarcelCoding) [reply java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:70]: Notifiy Users

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+        final Session session = sqlConnection.openSession();
+        session.beginTransaction();
+        session.save(channel);
+        session.getTransaction().commit();
+        session.close();
+        return channel;
+    }
+
+    public static void removeChannel(final UUID id) {
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java:41]: Notifiy Users

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public static Channel createChannel(final String name) {
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java:29]: Notifiy Users

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Message message = (Message) o;
+        return Objects.equals(getUser(), message.getUser()) &&
+                Objects.equals(getChannel(), message.getChannel()) &&
+                getType() == message.getType() &&
+                Objects.equals(getText(), message.getText()) &&
+                Objects.equals(getTarget(), message.getTarget());
+    }
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:129]: Add UUID

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(getUser(), getChannel(), getType(), getText(), getTarget());
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:133]: Add UUID

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+import org.hibernate.Session;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Table;
+import java.util.Objects;
+import java.util.UUID;
+
+@Entity
+@Table(name = "channel")
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java:15]: Use prefix 'chat_'

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
+package net.cryptic_game.backend.data.Chat;
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java:1]: Use lower case for package names

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
+package net.cryptic_game.backend.data.Chat;
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:1]: Use lower case package name

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+import javax.persistence.Entity;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+import java.util.List;
+import java.util.Objects;
+
+@Entity
+@Table(name = "channel_access")
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:19]: Add prefix 'chat_'

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+                    .setParameter("channel", channel)
+                    .setParameter("user", user);
+            return true;
+        } catch (HibernateException e) {
+            return false;
+        }
+    }
+
+    public static List<User> getMembers(final Channel channel) {
+
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:75]: Remove unnecessary empty line

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+import net.cryptic_game.backend.data.user.User;
+import org.hibernate.Session;
+import org.hibernate.annotations.Type;
+
+import javax.persistence.*;
+import java.util.List;
+import java.util.Objects;
+
+@Entity
+@Table(name = "message")
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:15]: Add prefix 'chat_'

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 18]
```diff
@@ -0,0 +1,147 @@
+package net.cryptic_game.backend.data.Chat;
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:1]: Use lower case for package names

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+    @Column(updatable = false)
+    private ChatAction type;
+
+    @Column(name = "text", updatable = false, nullable = true)
+    private String text;
+
+    /**
+     * user id of the target, if it's not a whisper its null and targets consequently
+     * everyone in the channel
+     */
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:38]: Change to: User ID of the target, if it is not a whisper it is null and consequently targets everyone in the channel

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+    public void setUser(User user) {
+        this.user = user;
+    }
+
+    public Channel getChannel() {
+        return channel;
+    }
+
+    public void setChannel(Channel channel) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:44]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+        this.channel = channel;
+    }
+
+    public static ChannelAccess join(User user, Channel channel) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:48]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+        channelAccess.setChannel(channel);
+
+        sqlSession.beginTransaction();
+        sqlSession.save(channelAccess);
+        sqlSession.getTransaction().commit();
+        sqlSession.close();
+        return channelAccess;
+    }
+
+    public static boolean leave(User user, Channel channel) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:62]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+    @OneToOne
+    @Column(name = "target", updatable = false, nullable = true)
+    private User target;
+
+    public User getUser() {
+        return user;
+    }
+
+    public void setUser(User user) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:47]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 24]
```diff
@@ -0,0 +1,147 @@
...
+        this.user = user;
+    }
+
+    public Channel getChannel() {
+        return channel;
+    }
+
+    public void setChannel(Channel channel) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:55]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+        this.channel = channel;
+    }
+
+    public ChatAction getType() {
+        return type;
+    }
+
+    public void setType(ChatAction type) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:63]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+        this.type = type;
+    }
+
+    public String getText() {
+        return text;
+    }
+
+    public void setText(String text) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:71]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+        this.text = text;
+    }
+
+    public User getTarget() {
+        return target;
+    }
+
+    public void setTarget(User target) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:79]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+        this.target = target;
+    }
+
+    public void send(Channel channel, User user, String text) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:83]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 29]
```diff
@@ -0,0 +1,147 @@
...
+        send(channel, user, ChatAction.SEND_MESSAGE, text);
+    }
+
+    public void send(Channel channel, User user, ChatAction type, String text) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:87]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+        send(channel, user, null, type, text);
+    }
+
+    public Message send(Channel channel, User user, User target, ChatAction type, String text) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:91]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+        message.setText(text);
+
+        sqlSession.beginTransaction();
+        sqlSession.save(message);
+        sqlSession.getTransaction().commit();
+        sqlSession.close();
+        return message;
+    }
+
+    public List<Message> getMessages(User user, Channel channel) {
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:108]: Parameters can be final

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public static void removeChannel(final UUID id) {
+        final Channel channel = getById(id);
+        if(channel != null) {
+            final Session session = sqlConnection.openSession();
+            session.beginTransaction();
+            session.delete(channel);
+            session.getTransaction().commit();
+            session.close();
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java:48]: Use delete() of superclass

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java [hunk 33]
```diff
@@ -0,0 +1,78 @@
...
+
+    public static Channel createChannel(final String name) {
+        final Channel channel = new Channel();
+        channel.setName(name);
+
+        final Session session = sqlConnection.openSession();
+        session.beginTransaction();
+        session.save(channel);
+        session.getTransaction().commit();
+        session.close();
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Channel.java:37]: Use saveOrUpdate() of superclass

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java [hunk 34]
```diff
@@ -0,0 +1,106 @@
...
+        Session sqlSession = sqlConnection.openSession();
+
+        ChannelAccess channelAccess = new ChannelAccess();
+        channelAccess.setUser(user);
+        channelAccess.setChannel(channel);
+
+        sqlSession.beginTransaction();
+        sqlSession.save(channelAccess);
+        sqlSession.getTransaction().commit();
+        sqlSession.close();
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/ChannelAccess.java:58]: Use saveOrUpdate() of superclass

File: java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java [hunk 35]
```diff
@@ -0,0 +1,147 @@
...
+        message.setUser(user);
+        message.setChannel(channel);
+        message.setTarget(target);
+        message.setType(type);
+        message.setText(text);
+
+        sqlSession.beginTransaction();
+        sqlSession.save(message);
+        sqlSession.getTransaction().commit();
+        sqlSession.close();
```
Comment (JannikEmmerich) [inline java-data/src/main/java/net/cryptic_game/backend/data/Chat/Message.java:104]: Use saveOrUpdate() of superclass

File: java-data/src/main/java/net/cryptic_game/backend/data/chat/ChannelAccess.java [hunk 36]
```diff
@@ -0,0 +1,111 @@
...
+
+    public static boolean leave(final User user, final Channel channel, Set<ApiClient> notifyUsers) {
+        try (Session sqlSession = sqlConnection.openSession()) {
+            sqlSession
+                    .createQuery("delete from ChannelAccess ca where ca.channel = :channel and ca.user = :user")
+                    .setParameter("channel", channel)
+                    .setParameter("user", user);
+            return true;
+        } catch (HibernateException e) {
+            final Logger log = LoggerFactory.getLogger(ChannelAccess.class);
```
Comment (MarcelCoding) [inline java-data/src/main/java/net/cryptic_game/backend/data/chat/ChannelAccess.java:72]: Define the Logger globaly as `private static final Logger log = LoggerFactory.getLogger(ChannelAccess.class);`

**Pull request merged**